share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorService.java

```diff
@@ -0,0 +1,536 @@
+        log.info("Startup complete.");
+    }
+
+    @Override
+    public void shutdown() {
+        if (!isActive.compareAndSet(true, false)) {
+            log.warn("Share coordinator is already shutting down.");
+            return;
+        }
+
+        log.info("Shutting down.");
+        Utils.closeQuietly(runtime, "coordinator runtime");
+        Utils.closeQuietly(shareCoordinatorMetrics, "share coordinator metrics");
+        log.info("Shutdown complete.");
+    }
+
+    @Override
+    public CompletableFuture<WriteShareGroupStateResponseData> writeState(RequestContext context, WriteShareGroupStateRequestData request) {
+        // Send an empty response if topic data is empty
+        if (isEmpty(request.topics())) {
+            log.error("Topic Data is empty: {}", request);
+            return CompletableFuture.completedFuture(
+                new WriteShareGroupStateResponseData()
+            );
+        }
+
+        // Send an empty response if partition data is empty for any topic
+        for (WriteShareGroupStateRequestData.WriteStateData topicData : request.topics()) {
+            if (isEmpty(topicData.partitions())) {
+                log.error("Partition Data for topic {} is empty: {}", topicData.topicId(), request);
+                return CompletableFuture.completedFuture(
+                    new WriteShareGroupStateResponseData()
+                );
+            }
+        }
+
+        String groupId = request.groupId();
+        // Send an empty response if groupId is invalid
+        if (isGroupIdEmpty(groupId)) {
+            log.error("Group id must be specified and non-empty: {}", request);
+            return CompletableFuture.completedFuture(
+                new WriteShareGroupStateResponseData()
+            );
+        }
+
+        // Send an empty response if the coordinator is not active
+        if (!isActive.get()) {
+            return CompletableFuture.completedFuture(
+                generateErrorWriteStateResponse(
+                    request,
+                    Errors.COORDINATOR_NOT_AVAILABLE,
+                    "Share coordinator is not available."
+                )
+            );
+        }
+
+        // The request received here could have multiple keys of structure group:topic:partition. However,
+        // the writeState method in ShareCoordinatorShard expects a single key in the request. Hence, we will
+        // be looping over the keys below and constructing new WriteShareGroupStateRequestData objects to pass
+        // onto the shard method.
+        Map<Uuid, Map<Integer, CompletableFuture<WriteShareGroupStateResponseData>>> futureMap = new HashMap<>();
+        long startTime = time.hiResClockMs();
+
+        request.topics().forEach(topicData -> {
+            Map<Integer, CompletableFuture<WriteShareGroupStateResponseData>> partitionFut =
+                futureMap.computeIfAbsent(topicData.topicId(), k -> new HashMap<>());
+            topicData.partitions().forEach(
```

Comment:
Personally, I find the logic here a little hard to follow with the nesting and futures. I'd like us to consider refactoring this a bit in a future PR.